src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php

```diff
@@ -258,6 +262,67 @@ public function downloadVirtualFileAction(int $virtualProductFileId): BinaryFile
         return $response;
     }
 
+    /**
+     * @AdminSecurity("is_granted(['read'], request.get('_legacy_controller'))")
+     *
+     * @param Request $request
+     * @param string $languageCode
+     *
+     * @return JsonResponse
+     *
+     * @throws ProductConstraintException
+     */
+    public function getAjaxProductsAction(Request $request, string $languageCode): JsonResponse
+    {
+        /** @var LangRepository $langRepository */
+        $langRepository = $this->get('prestashop.core.admin.lang.repository');
+        $lang = $langRepository->getOneByLocaleOrIsoCode($languageCode);
+
+        $shopContext = $this->get('prestashop.adapter.shop.context');
+        $shopId = (int) $shopContext->getContextShopID();
+        if (empty($shopId)) {
+            $configuration = $this->get('prestashop.adapter.legacy.configuration');
+            $shopId = $configuration->getInt('PS_SHOP_DEFAULT');
+        }
+
+        try {
+            /** @var ProductForAssociation[] $products */
+            $products = $this->getQueryBus()->handle(new SearchProductsForAssociation(
+                $request->get('query', ''),
+                $lang->getId(),
+                $shopId
+            ));
+        } catch (ProductConstraintException $e) {
+            return $this->json([
+                'success' => false,
+                'message' => $e->getMessage(),
+            ], Response::HTTP_BAD_REQUEST);
+        }
+
+        if (empty($products)) {
+            return $this->json([], Response::HTTP_NOT_FOUND);
+        }
+
+        $productsData = [];
+        foreach ($products as $product) {
+            $productName = $product->getName();
+            if (!empty($product->getReference())) {
+                $productName .= sprintf(
+                    ' (ref: %s)',
+                    $product->getReference()
+                );
+            }
+
+            $productsData[] = [
+                'id' => $product->getProductId(),
+                'name' => $productName,
+                'image' => $product->getImageUrl(),
+            ];
+        }
+
+        return $this->json($productsData);
```

Comment:
In order to make the controller easier to read, may I suggest
```
return $this->json($this->formatFoundProductData($productsData));
```

Reply:
I thought about this as well, but was not sure it would bring that much more readability, at least the whole controller code is readable in a single function here, including its formatting. The controller itself is relatively short, around 60lines most of them being the validation of inputs and json error formatting.

If an effort was to be done to reduce this controller, I'd invest more time into a global `$this->jsonError` method that would allow returning an error response that would be common to all our APIs, for now each one handles its error output and error messages the way it wants to without any convention.

Such method would both reduce size and bring a more harmonized convention in our APIs.

Reply:
> I thought about this as well, but was not sure it would bring that much more readability

It does 😉  because it helps the reader. The reader, when he sees it, can tell himself "oh, OK, this function does the formatting, but because I'm trying to check how the data is fetched, not formatted, I don't need to open this function and see what is inside".

As written [here](https://softwareengineering.stackexchange.com/questions/132019/what-is-the-value-in-hiding-the-details-through-abstractions-isnt-there-value), "there is the other major "benefit" of abstraction, which is to hide implementation logic and details from users of this abstraction. The argument is that you don't need to know the details, and that one should concentrate on their own logic at this point."

If you read your controller action out loud and, you'll find that relying on `formatFoundProductData` allows you to clearly voice the intent/objective of this part of the code.

In the same idea, it's OK to do
```
if ($a->isDivisibleBy($b)) {
```
with
```
private function isDivisibleBy(int $x)
{
    return ($this % $x) === 0;
}
```

Reply:
Ok, I'm convinced ^^ I added a formatting function